R/pcair.R

```diff
@@ -1,299 +1,133 @@
+        )
+        samp.load <- suppressMessages(
+            snpgdsPCASampLoading(snp.load, gdsobj = gdsobj,
+                                 sample.id = part$rels,
+                                 num.thread = num.thread, verbose = verbose)
+        )
 
-            # load genotype data
-            geno <- getGenotypeSelection(genoData, snp = snp.include$index[snp.block.idx], scan = scan.include$index, drop = FALSE)
+        # combine unrelated and related PCs and order as in GDS file
+        eigenvect <- rbind(pca.unrel$eigenvect, samp.load$eigenvect)
+        rownames(eigenvect) <- c(as.character(pca.unrel$sample.id),
+                                 as.character(samp.load$sample.id))
+        sample.id <- as.character(read.gdsn(index.gdsn(gdsobj, "sample.id")))
+        sample.id <- intersect(sample.id, rownames(eigenvect))
+        samp.ord <- match(sample.id, rownames(eigenvect))
+        eigenvect <- eigenvect[samp.ord,]
 
-            # allele freq est from entire sample
-            pA <- 0.5*rowMeans(geno, na.rm = TRUE)
-            # remove monomorphic SNPs
-            snp.excl <- which(is.na(pA) | pA <= MAF | pA >= (1-MAF))
-            if(length(snp.excl > 0)){
-                geno <- geno[-snp.excl,]
-                pA <- pA[-snp.excl]
-            }
-            nsnps <- nsnps + length(pA)
-            
-            # Standardized genotype values
-            # estimated variance at each SNP
-            sigma.hat <- sqrt(2*pA*(1-pA))
-            # z_i = (x_i - 2\hat{p})/sqrt{2*\hat{p}(1-\hat{p})}
-            Z <- (geno-2*pA)/sigma.hat
-            Z[which(is.na(Z))] <- 0
-            
-            # empirical correlation matrix
-            Psi <- Psi + crossprod(Z)
-        }
-        Psi <- (1/nsnps)*Psi
-        
-        # sPCA analysis
-        if(verbose){  message("Performing Standard PCA...")  }
-        eig <- eigen(Psi, symmetric=TRUE)
-        
-        # subset desired number of eigenvectors
-        if(is.null(v)){	v <- scan.include$n }
-        # output
-        EIG <- eig$vectors[,1:v]
-        L <- eig$values[1:v]
-        sum.values <- sum(eig$values)
+    } else {
+        method <- "Standard PCA"
+        eigenvect <- pca.unrel$eigenvect
     }
+    eigenval <- pca.unrel$eigenval[1:ncol(pca.unrel$eigenvect)]
     
-    # add scanIDs as rownames of EIG
-    rownames(EIG) <- scan.include$value
-    
-    # return results
-    out <- list(vectors = EIG, 
-                values = L, 
-                sum.values = sum.values, 
-                rels = rels, 
-                unrels = unrels,
+    # output object
+    out <- list(vectors = eigenvect,
+                values = eigenval,
+                rels = part$rels,
+                unrels = part$unrels,
                 kin.thresh = kin.thresh,
-                div.thresh = -abs(div.thresh),
-                nsamp = scan.include$n,
-                nsnps = nsnps,
-                MAF = MAF,
+                div.thresh = div.thresh,
```

Comment:
I'm not enforcing div.thresh to be negative in the function anymore, so this doesn't need to be wrapped in -abs()